app/src/main/java/com/infomaniak/drive/ui/home/HomeViewModel.kt

```diff
@@ -18,78 +18,121 @@
 package com.infomaniak.drive.ui.home
 
 import androidx.collection.arrayMapOf
-import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
-import androidx.lifecycle.liveData
+import androidx.lifecycle.viewModelScope
 import com.infomaniak.drive.data.api.ApiRepository
 import com.infomaniak.drive.data.cache.FileController
 import com.infomaniak.drive.data.models.FileActivity
 import com.infomaniak.lib.core.models.ApiResponse
 import com.infomaniak.lib.core.models.ApiResponseStatus.SUCCESS
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.ensureActive
+import kotlinx.coroutines.launch
 import java.util.Date
 
 class HomeViewModel : ViewModel() {
-    private var lastActivityJob = Job()
 
-    var lastActivityPage = 1
+    var lastActivitiesResult = MutableLiveData<LastActivityResult?>()
+
+    private var lastActivityJob: Job? = null
 
-    var lastActivityLastPage = 1
     private var lastActivitiesTime: Long = 0
+    private var currentCursor: String? = null
     private var lastActivities = arrayListOf<FileActivity>()
     private var lastMergedActivities = arrayListOf<FileActivity>()
 
-    fun getLastActivities(
+    fun loadLastActivities(driveId: Int, forceDownload: Boolean = false) {
+        loadLastActivities(driveId, forceDownload, isFirstPage = true)
+    }
+
+    fun loadMoreActivities(driveId: Int) {
+        loadLastActivities(driveId, forceDownload = false, cursor = currentCursor)
+    }
+
+    fun restoreActivitiesIfNeeded() {
+        if (lastMergedActivities.isNotEmpty()) {
+            lastActivitiesResult.value = LastActivityResult(
+                mergedActivities = lastMergedActivities,
+                isComplete = currentCursor != null,
+                isFirstPage = true,
+            )
+        }
+    }
+
+    private fun loadLastActivities(
         driveId: Int,
-        forceDownload: Boolean = false
-    ): LiveData<Pair<ApiResponse<ArrayList<FileActivity>>, ArrayList<FileActivity>>?> {
-        lastActivityJob.cancel()
-        lastActivityJob = Job()
+        forceDownload: Boolean = false,
+        isFirstPage: Boolean = false,
+        cursor: String? = null,
+    ) {
+        lastActivityJob?.cancel()
+        lastActivityJob = viewModelScope.launch(Dispatchers.IO) {
+            val lastActivityResult = getLastActivities(driveId, forceDownload, isFirstPage, cursor)
+            if (lastActivityJob?.isActive == true) lastActivitiesResult.postValue(lastActivityResult)
```

Comment:
What is the difference with ensureActive() here ?

Reply:
`ensureActive` throw a CancellationException

Reply:
Edit: In this case, both have the same effect, but as the code is small and easy too read, isActive() was good here
(ensureActive() would need an extra ligne)